File: backend/python_server/preprocessing.py
import cv2

# ★★★ スマートリサイズ関数を新しく追加 ★★★
def smart_resize(image, max_dim=2048):
    """
    画像の縦横比を維持したまま、長辺がmax_dimになるようにリサイズする
    """
    h, w, _ = image.shape
    # すでに十分に小さい場合は何もしない
    if h <= max_dim and w <= max_dim:
        return image

    # 長辺がどちらかを判断
    if w > h:
        new_w = max_dim
        new_h = int(h * (max_dim / w))
    else:
        new_h = max_dim
        new_w = int(w * (max_dim / h))
    
    # cv2.INTER_AREAは、画像を縮小する際に最も品質が良いとされる補間方法
    return cv2.resize(image, (new_w, new_h), interpolation=cv2.INTER_AREA)

File: backend/rust_axum_server/src/preprocessing.rs
use image::{DynamicImage, imageops::FilterType};

pub fn smart_resize(image: &DynamicImage, max_dim: u32) -> DynamicImage {
    let (w, h) = (image.width(), image.height());

    if w <= max_dim && h <= max_dim {
        return image.clone();
    }

    let (new_w, new_h) = if w > h {
        let new_w = max_dim;
        let new_h = (h as f64 * (max_dim as f64 / w as f64)) as u32;
        (new_w, new_h)
    } else {
        let new_h = max_dim;
        let new_w = (w as f64 * (max_dim as f64 / h as f64)) as u32;
        (new_w, new_h)
    };

    image.resize(new_w, new_h, FilterType::Lanczos3)
}
